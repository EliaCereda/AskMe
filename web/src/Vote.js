import React, {Component} from "react";

class Vote extends Component {

    handleClick(value) {
        this.props.onVote({ value,
                            user: this.props.currentUser,
                            answer_id: this.props.answerId });
    }

    formattedScore() {
        const score = this.props.score;

        if (score > 0) {
            return <span>+{score}</span>;
        } else if (score === 0) {
            return <span>&nbsp;&nbsp;0</span>;
        } else {
            return <span>−{-score}</span>;
        }
    }

    render() {
        const isLoggedIn = (this.props.currentUser !== null);

        let upvote;
        let downvote;

        if (isLoggedIn) {
            const vote = this.props.votes.find((v) => v.user === this.props.currentUser);

            if (vote) {
                upvote = (vote.value === +1) ? 'selected' : '';
                downvote = (vote.value === -1) ? 'selected' : '';
            }
        }

        return (
            <div className="Vote">
                <button className={upvote} disabled={!isLoggedIn} onClick={() => this.handleClick(+1)}>⬆︎</button>
                <button className={downvote} disabled={!isLoggedIn} onClick={() => this.handleClick(-1)}>⬇︎</button>

                {this.formattedScore()} votes
            </div>
        );
    }
}

export default Vote;
