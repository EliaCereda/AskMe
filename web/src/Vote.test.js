import React from 'react';
import Vote from "./Vote";

describe('CreateQuestion', () => {
    it('renders correctly', () => {
        let wrapper;

        wrapper = shallow(<Vote currentUser="Giorgia" score={2} votes={[]}/>);
        expect(wrapper).toMatchSnapshot();

        let votes;

        votes = [
            {
                "_links": {},
                "answer_id": 2,
                "user": "Giorgia",
                "value": -1
            },
            {
                "_links": {},
                "answer_id": 2,
                "user": "fghj",
                "value": 1
            }
        ];

        wrapper = shallow(<Vote currentUser="Giorgia" score={2} votes={votes}/>);
        expect(wrapper).toMatchSnapshot();

        votes = [
            {
                "_links": {},
                "answer_id": 2,
                "user": "asd",
                "value": -1
            },
            {
                "_links": {},
                "answer_id": 2,
                "user": "fghj",
                "value": 1
            }
        ];

        wrapper = shallow(<Vote currentUser="Giorgia" score={2} votes={votes}/>);
        expect(wrapper).toMatchSnapshot();

        votes = [
            {
                "_links": {},
                "answer_id": 2,
                "user": "Giorgia",
                "value": +1
            },
            {
                "_links": {},
                "answer_id": 2,
                "user": "fghj",
                "value": 1
            }
        ];

        wrapper = shallow(<Vote currentUser="Giorgia" score={2} votes={votes}/>);
        expect(wrapper).toMatchSnapshot();

        wrapper = shallow(<Vote currentUser={null} score={2}/>);
        expect(wrapper).toMatchSnapshot();
    });
});
