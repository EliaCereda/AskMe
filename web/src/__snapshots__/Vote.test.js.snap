// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CreateQuestion renders correctly 1`] = `
<div
  className="Vote"
>
  <button
    disabled={false}
    onClick={[Function]}
  >
    ⬆︎
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    ⬇︎
  </button>
  2
   votes
</div>
`;

exports[`CreateQuestion renders correctly 2`] = `
<div
  className="Vote"
>
  <button
    className=""
    disabled={false}
    onClick={[Function]}
  >
    ⬆︎
  </button>
  <button
    className="selected"
    disabled={false}
    onClick={[Function]}
  >
    ⬇︎
  </button>
  2
   votes
</div>
`;

exports[`CreateQuestion renders correctly 3`] = `
<div
  className="Vote"
>
  <button
    disabled={false}
    onClick={[Function]}
  >
    ⬆︎
  </button>
  <button
    disabled={false}
    onClick={[Function]}
  >
    ⬇︎
  </button>
  2
   votes
</div>
`;

exports[`CreateQuestion renders correctly 4`] = `
<div
  className="Vote"
>
  <button
    className="selected"
    disabled={false}
    onClick={[Function]}
  >
    ⬆︎
  </button>
  <button
    className=""
    disabled={false}
    onClick={[Function]}
  >
    ⬇︎
  </button>
  2
   votes
</div>
`;

exports[`CreateQuestion renders correctly 5`] = `
<div
  className="Vote"
>
  <button
    disabled={true}
    onClick={[Function]}
  >
    ⬆︎
  </button>
  <button
    disabled={true}
    onClick={[Function]}
  >
    ⬇︎
  </button>
  2
   votes
</div>
`;
